File: 2015/16/16.rs
aoc::main!();

fn parser(input: &str) -> Vec<HashMap<&str, u16>> {
    input
    .lines()
    .map(|l| -> Vec<&str> {l.split(&[':',',',' ']).collect_vec()})
    .map(|v| 
        HashMap::from([
            (v[3], v[5].parse().unwrap()),
            (v[7],  v[9].parse().unwrap()),
            (v[11], v[13].parse().unwrap())]))
    .collect_vec()
}


fn part_1(aunts : Vec<HashMap<&str, u16>>) -> usize {
    find_sue(aunts, Vec::new(), Vec::new())
}

fn part_2(aunts : Vec<HashMap<&str, u16>>) -> usize {
    find_sue(aunts, Vec::from(["cats", "trees"]), Vec::from(["pomeranians", "goldfish"]))
}

fn find_sue(aunts : Vec<HashMap<&str, u16>>, lower: Vec<&str>, greater: Vec<&str>) -> usize {
    let sue = HashMap::from([
        ("akitas", 0)  , ("children", 3)   ,  ("cars", 2)    ,  ("cats", 7) , ("goldfish", 5), 
        ("perfumes", 1), ("pomeranians", 3),  ("samoyeds", 2),  ("trees", 3), ("vizslas", 0)]);
    
    aunts
    .iter()
    .find_position(|aunt| 
        aunt
        .iter()
        .all(|(k, v)| 
            if lower.contains(k) {sue.get(k).unwrap() < v}
            else if greater.contains(k) {sue.get(k).unwrap() > v} 
            else {sue.get(k).unwrap() == v}))
    .unwrap()
    .0 
    + 1
    }

File: 2015/16/16.py
from parse    import parse
from operator import eq, lt, gt
from copy     import deepcopy

def parser(data: str): 
    aunts: dict = {}
    for aunt_data in data.splitlines():
        n, c1, v1, c2, v2, c3, v3 = parse("Sue {:d}: {}: {:d}, {}: {:d}, {}: {:d}", aunt_data)
        aunts[n] = {c1: v1, c2: v2, c3: v3}
    return aunts

def solver(aunts: dict):
    yield find_Sue(deepcopy(aunts), eq, eq)
    yield find_Sue(aunts, gt, lt)
        
def find_Sue(aunts: dict, op1, op2):
    Sue: dict = { "children": 3, "cats":     7, "samoyeds": 2, "pomeranians":  3, "akitas":   0, \
                  "vizslas":  0, "goldfish": 5, "trees":    3, "cars":         2, "perfumes": 1}
    
    for n, compounds in aunts.items():
        for compound, value in list(compounds.items()):
            if compound in ['cats', 'trees']:
                if op1(value, Sue[compound]): del aunts[n][compound]
                else: break
            elif compound in ['pomeranians', 'goldfish']:
                if op2(value, Sue[compound]): del aunts[n][compound]
                else: break
            elif value == Sue[compound]: del aunts[n][compound]
        if compounds == {}: return n